spec/index.bs

```diff
@@ -569,6 +561,63 @@ enum GPUBindingType {
+    The dimensions and multi-sampling properties of texture bindings.
+
+    Note: This allows Metal-based implementations to back the respective bind
+    groups with `MTLArgumentBuffer` objects that are more efficient to bind at
+    run-time.
+
+  * {{GPUBindGroupLayoutBinding/dynamic}}:
+    For [=uniform-buffer=], [=storage-buffer=], and [=readonly-storage-buffer=] bindings, 
+    indicates that the binding has a dynamic offset. One offset must be passed to 
+    setBindGroup for each dynamic binding in increasing order of
+    {{GPUBindGroupLayoutBinding/binding}} number.
+
+### {{GPUDevice/createBindGroupLayout()|GPUDevice.createBindGroupLayout(GPUBindGroupLayoutDescriptor)}} ### {#GPUDevice-createBindGroupLayout}
+
+A valid {{GPUBindGroupLayout}} object has an <dfn>internal map</dfn> of |u32| <dfn>binding number</dfn>s to internal <dfn>layout binding</dfn> objects. 
+Each [=layout binding=] has a {{GPUShaderStageFlags}} <dfn>visibility</dfn> bitset and a {{GPUBindingType}} <dfn>type</dfn>.
+
+A [=layout binding=] with type {{GPUBindingType/uniform-buffer}}, {{GPUBindingType/storage-buffer}}, or {{GPUBindingType/readonly-storage-buffer}} also has 
+a <dfn>dynamic</dfn> boolean which must be initially <code>false</code>.
+
+A [=layout binding=] with type {{GPUBindingType/sampled-texture}} or {{GPUBindingType/storage-texture}} also has a {{GPUTextureViewDimension}} <dfn>textureDimension</dfn> which must be initially <code>"2d"</code>,
+a {{GPUTextureComponentType}} <dfn>textureComponentType</dfn> which must be initially <code>"float</code>, and a <dfn>multisampled</dfn> boolean which must be initially <code>false</code>.
+
+<div algorithm="GPUDevice.createBindGroupLayout(descriptor)">
+Any time during this algorithm, to <dfn>return with error</dfn> is to:
+    1. Generate a validation error in the current scope with appropriate error message.
+    1. Create a new invalid {{GPUBindGroupLayout}} and return the result.
+
+The <dfn method for "GPUDevice">createBindGroupLayout(|descriptor|)</dfn> method attempts to create a valid {{GPUBindGroupLayout}}.
+
+1. If the device is [=lost=], [=return with error=].
+1. Let |layout| be a new valid {{GPUBindGroupLayout}} object.
+1. For each {{GPUBindGroupLayoutBinding}} |bindingDescriptor| in |descriptor|.{{GPUBindGroupLayoutDescriptor/bindings}}:
+    1. If |bindingDescriptor|.{{GPUBindGroupLayoutBinding/binding}} is already present in |layout|'s [=internal map=], [=return with error=].
+    1. Let |layoutBinding| be a new internal [=layout binding=] object.
+    1. If |bindingDescriptor|.{{GPUBindGroupLayoutBinding/type}} is {{GPUBindingType/uniform-buffer}}:
+        1. If |layout|'s [=internal map=] contains {{GPULimits/maxUniformBuffersPerShaderStage|GPULimits.maxUniformBuffersPerShaderStage}} [=layout binding=]s of [=type=] {{GPUBindingType/uniform-buffer}} that all share any [=visibility=] bit contained in |bindingDescriptor|.{{GPUBindGroupLayoutBinding/visibility}}, [=return with error=].
```

Comment:
I'd add these validations at the end of the algorithm even if that would be allowed by implementations under the as-if rule. Basically this makes validation O(n2) instead of O(n)

Reply:
+1, I think it would be much easier to read too.

Reply:
it's hard to read those long lines, I'll be happy to take another look once this is refactored according to this suggestion

Reply:
I've just finished refactoring a bunch of stuff. PTAL!